drivers/usbdev_synopsys_dwc2/usbdev_synopsys_dwc2.c

```diff
@@ -750,12 +756,78 @@ static void _usbdev_init(usbdev_t *dev)
             /* disable ULPI FS/LS serial interface */
             _global_regs(usbdev->config)->GUSBCFG &= ~USB_OTG_GUSBCFG_ULPIFSLS;
         }
+
+#elif defined(MODULE_PERIPH_USBDEV_HS_UTMI)
+        else if (conf->phy == DWC2_USB_OTG_PHY_UTMI) {
+            /* enable ULPI clock */
+            periph_clk_en(conf->ahb, RCC_AHB1ENR_OTGHSULPIEN);
+            /* enable UTMI HS PHY Controller clock */
+            periph_clk_en(APB2, RCC_APB2ENR_OTGPHYCEN);
+
+#ifdef USB_OTG_GUSBCFG_ULPI_UTMI_SEL
+            /* select UTMI+ PHY */
+            _global_regs(usbdev->config)->GUSBCFG &= ~USB_OTG_GUSBCFG_ULPI_UTMI_SEL;
+#endif /* USB_OTG_GUSBCFG_ULPI_UTMI_SEL */
+#ifdef USB_OTG_GUSBCFG_PHYIF
+            /* use the 8-bit interface and single data rate */
+            _global_regs(usbdev->config)->GUSBCFG &= ~USB_OTG_GUSBCFG_PHYIF;
+#endif /* USB_OTG_GUSBCFG_PHYIF */
+
+            /* disable the on-chip FS transceiver */
+            _global_regs(usbdev->config)->GUSBCFG &= ~USB_OTG_GUSBCFG_PHYSEL;
+
+            /* configure the USB HS PHY Controller (USB_HS_PHYC),
+             * USB_HS_PHYC and GCCFG are STM32 specific */
+#ifdef USB_HS_PHYC
+            /* enable USB HS PHY Controller */
+            _global_regs(usbdev->config)->GCCFG |= USB_OTG_GCCFG_PHYHSEN;
+
+            /* determine the PLL input clock of the USB HS PHY from HSE clock */
+            switch (CLOCK_HSE) {
+                case 12000000:
+                    USB_HS_PHYC->USB_HS_PHYC_PLL |= USB_HS_PHYC_PLL1_PLLSEL_12MHZ;
+                    break;
+                case 12500000:
+                    USB_HS_PHYC->USB_HS_PHYC_PLL |= USB_HS_PHYC_PLL1_PLLSEL_12_5MHZ;
+                    break;
+                case 16000000:
+                    USB_HS_PHYC->USB_HS_PHYC_PLL |= USB_HS_PHYC_PLL1_PLLSEL_16MHZ;
+                    break;
+                case 24000000:
+                    USB_HS_PHYC->USB_HS_PHYC_PLL |= USB_HS_PHYC_PLL1_PLLSEL_24MHZ;
+                    break;
+                case 25000000:
+                    USB_HS_PHYC->USB_HS_PHYC_PLL |= USB_HS_PHYC_PLL1_PLLSEL_25MHZ;
+                    break;
+                default:
+                    assert(0);
+            }
+
+            /* configure the tuning interface of the USB HS PHY,
+             * the magic value (USB_HS_PHYC_TUNE_VALUE) from STM32CubeF7 HAL
+             * Driver MCU Component for F7 is used */
+            USB_HS_PHYC->USB_HS_PHYC_TUNE |= 0x00000F13U;
```

Comment:
Based on the stm32f723 reference manual, this register doesn't look that magical. It does seem to contain quite a number of board-dependent options. Does it make sense to add this to the configuration struct together with some generic 'recommended' value (this `0x0F13U` maybe)?

Reply:
> Based on the stm32f723 reference manual, this register doesn't look that magical.

Yes, I have seen it. However, the number of possible combinations of these options is quite large and might lead to misconfigurations. To be honest, I wouldn't have any idea how to configure it properly. The value is taken from [STM32CubeF7](https://github.com/STMicroelectronics/STM32CubeF7/blob/f8bda023e34ce9935cb4efb9d1c299860137b6f3/Drivers/STM32F7xx_HAL_Driver/Inc/stm32f7xx_ll_usb.h#L276-L278).

We either add it to the configurationstructure or we could use something like the STM32CubeF7 HAL driver in `drivers/include/usbdev_synopsys_dwc2.h`
```c
#ifndef USB_HS_PHYC_TUNE_VALUE
#define USB_HS_PHYC_TUNE_VALUE     0x00000F13U
#endif
``` 
and use this define in the driver. The value can then be overridden in the board configuration.

Reply:
> We either add it to the configuration structure or we could use something like the STM32CubeF7 HAL driver in `drivers/include/usbdev_synopsys_dwc2.h`

I'm fine with either solution, I don't think we need to have a field in the config struct for every individual bit in that register though.